0011-command-execution-protocol/README.md

```diff
@@ -0,0 +1,175 @@
+- An endpoint may support multiple command types, thus it must tell them apart.
+- Command may success or fail. The caller must know outcomes of the command or the reason of a failure.
+- Commands are often not idempotent. That is, executing a single command twice may result in a different (possibly undesired) outcome.
+- An endpoint may be offline at the time of command invocation. The caller must be informed that the command execution did not succeed.
+
+# Use cases
+## UC1: Remote control
+An endpoint operates a device. Command Execution Protocol may be used to transfer commands to the endpoint and instruct it to change device behavior.
+
+## UC2: Expose local resources
+An endpoint represents a sensor that does not measure the value periodically. Command Execution Protocol may be used to issue a measurement command to the endpoint and transfer measurement results back to the caller.
+
+## UC3: OTA update
+> Note: while OTA updates require more support and probably a separate protocol, a basic version can be implemented using command execution.
+
+An endpoint may receive a command to update its firmware. When update is complete, the endpoint sends an update result to the server.
+
+# Design
+The Command Execution Protocol extension consists of two resource sets with the corresponding resource paths:
+- `/<endpoint_token>/command/<command_type>` - command resource
+- `/<endpoint_token>/result/<command_type>` - result resource
+
+## Command resource
+Command resource is used to distribute commands to the endpoints.
+
+The command resource is a request/response resource with an observe extension which resides at the following resource path:
+```
+/<endpoint_token>/command/<command_type>
+```
+where `<command_type>` is a type of the command the endpoint wants to serve.
+
+For MQTT, observe is achieved by publishing multiple responses to the response topic; for CoAP, it is achieved with the `Observe` option defined in [RFC 7641](https://tools.ietf.org/html/rfc7641).
+
+Command type MUST be a non-empty alpha-numeric string identifying the command type to the endpoint.
+
+### Command request
+The endpoint command requests MUST be a UTF-8 encoded JSON object with the JSON Schema defined in [command-request.schema.json](./command-request.schema.json) file.
+
+```json
+{
+  "$schema": "http://json-schema.org/schema#",
+  "title": "11/CEP command request schema",
+
+  "type": "object",
+  "properties": {
+    "observe": {
+      "type": "boolean",
+      "description": "Whether to send upcoming command for this endpoint."
+    }
+  },
+  "additionalProperties": false
+}
+```
+
+If `observe` field is present and is `true`, the server MUST send new commands to the endpoint whenever they arrive.
+
+If `observe` field is present and is `false`, the server MUST NOT send new commands to the endpoint whenever they arrive.
```

Comment:
It would be hard to implement especially if one endpoint will send `observe=false` in scope of one client session and `observe=true` in scope of another.